### ISSUE TITLE
Define an architecture for authentication with no library dependency

### ISSUE THREAD

**Issue by rodolfomiranda** [WebOfTrust/signify-browser-extension#23]
The goal is to have a way to authenticate requests from web pages without the need to modify or add extra code to the webpage html. Since Manifest v3 does not allow to dynamically change header during the network request, signatures should be created in advance and will act as a temporal auth token. We need to investigate further.

**Comment by rodolfomiranda** [comment]
We reviewing the requirement of this issue by loosening the "no library dependency" to "adoption with minimum changes" but keeping the non-interactive approach of signing headers with timestamp and origin url as part of the signature.

A way to minimize the changes needed by webpage developers to adopt keri auth headers is to implement [monkey patching](https://en.wikipedia.org/wiki/Monkey_patch) that is a technique to modify or override a function at runtime. This means that a webpage only needs to execute a simple code at start up  to monkey-patch the ‘fetch’ or the `XMLHttpRequest` function to create an interceptor of all network requests. For example, to add auth headers to all request performed by the javascript  function `Fetch` we can do something like
```
window.fetch = new Proxy(window.fetch, {
  apply: function (target, that, args) {
    const authHeaders = extensionAPI.getAuthHeaders();
    args[1].headers = {…args[1].headers, …authHeaders}
    let temp = target.apply(that, args);
    temp.then((res) => {
      // After completion of request
      // Verify response headers
    });
    return temp;
  },
});
```

Something similar can be done for XMLHttpRequest which is the older browser function used by Ajax and by the AXIOS library. Additionally, AXIOS provides a built in interceptor that can also be used depending on how the page is implemented. This type of  interceptors can also be tuned to add headers only in specific conditions if needed. Note that this won’t work for resources loaded from HTML tags directly such as ‘img' or ‘iframe' where you specify the URL (I don’t know if  there’s a way to add headers unless you make the request from a Fetch or XMLHttpRequest)

We are thinking of providing a guide with libraries, code snippets and recommendations that can help webpage implementors to adopt this authentication method with minimum effort for most common webpage architectures and frameworks. 

In theory, the monkey patch script can be injected from the web extension as well and should work seamlessly. However, I learned that some web pages already implement techniques to prevent the injection and that extension with that sort of code may be considered untrustworthy and not be  allowed in the chrome store. That’s why we discarded that option initially. However,  we are now reconsidering the idea knowing that webpage maintainers are actually opting for the new authentication method and may be willing to make small code additions.
